### PR TITLE
TINY-6732: Drop mocha and swap to implementing a compatible interface

### PR DESCRIPTION
**Pull request by lnewson**
Mocha turned out to be problematic as it requires multiple polyfills to run on IE. This meant that some tests may be passing due to the polyfills it bundles as part of its web package, as the functionality wouldn't normally be there (it bundles most core-js polyfills from what I could tell).

As such, we decided before Christmas to revert back to implementing a compatible interface instead. This also adds one minor fix/improvement to normalize the assertion errors.

Note: I've also tested this by converting all the core TinyMCE tests (about 235 files).

### CODE REVIEW

File: modules/runner/src/test/ts/runner/HooksTest.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { HookType, Suite } from '@ephox/bedrock-common';
+import { assert } from 'chai';
+import { describe, it } from 'mocha';
```
Comment (JamesToohey) [inline modules/runner/src/test/ts/runner/HooksTest.ts:3]: `mocha`?

**Review by JamesToohey — COMMENTED** [review]
I have a few questions about some things that we can go through in the new year. But looks good 👍

File: modules/runner/src/main/ts/core/Globals.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+const root = createSuite('');
+const suiteStack: Suite[] = [ root ];
+
+const getCurrentSuite = (): Suite | undefined =>
+  suiteStack.length > 0 ? suiteStack[suiteStack.length - 1] : undefined;
+
+const getCurrentSuiteOrDie = (): Suite => {
+  const currentSuite = getCurrentSuite();
+  if (currentSuite === undefined) {
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Globals.ts:15]: Katamari can't be used in bedrock, as it's the lowest item in our dependency tree (ie katamari relies on bedrock).

File: modules/runner/src/main/ts/core/Suite.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      [HookType.After]: [],
+      [HookType.AfterEach]: [],
+    },
+    suites: [],
+    tests: [],
+    parent,
+    _only: false,
+    _skip: false,
+    isSkipped: () => {
+      return suite._skip || parent?.isSkipped() === true;
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Suite.ts:18]: You can't do this, as the return type then would be `undefined | boolean` due to the optional chaining here.

File: modules/runner/src/main/ts/core/Test.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { Suite, ExecuteFn, Test } from '@ephox/bedrock-common';
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Test.ts:1]: Not sure what you mean here sorry? I can't see anything the same, as they have minor differences.

File: modules/runner/src/main/ts/runner/Hooks.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+  } else {
+    const hooks: Hook[] = suite.hooks[type];
+    if (parents && suite.parent !== undefined) {
+      const allHooks = hooks.concat(getHooks(suite.parent, type, parents));
+      return type === HookType.AfterEach || type === HookType.After ? allHooks.reverse() : allHooks;
+    } else {
+      return hooks;
+    }
+  }
+};
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: `runOnParents` here doesn't make sense, as you aren't running anything.

File: modules/runner/src/test/ts/runner/HooksTest.ts (hunk 1 above)
Comment (lnewson) [reply modules/runner/src/test/ts/runner/HooksTest.ts:3]: Mocha is what's used to test the bedrock logic, so this is expected.

File: modules/runner/src/main/ts/core/Globals.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+export const it = (title: string, fn: ExecuteFn): Test => {
+  const suite = getCurrentSuiteOrDie();
+  const test = createTest(title, fn, suite);
+
+  suite.tests.push(test);
+  Register.test(test);
+
+  return test;
+};
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Globals.ts:77]: Fixed. I specifically didn't want to space out `root` and `suiteStack` as those are variables, not functions.

File: modules/runner/src/main/ts/runner/Hooks.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: That's what it's called when it's passed to it on line 37?

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: Yes, but that function is called `runHooks` so it's executing the hooks, not fetching them.

[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: `getParents` then?

File: modules/runner/src/main/ts/core/Suite.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/core/Suite.ts:18]: Sorry I meant to rearrange (I swear I did!?)

```suggestion
      return parent?.isSkipped() || suite._skip;
```

Happy to leave as is though, it's tiny(tm)

File: modules/runner/src/main/ts/runner/Hooks.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: It's just that `parents` implies a `boolean` as to whether or not it has them, not whether or not it should get them

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: `getParents` is a very bad name for a variable, as it implies it's a getter function. I'll see what else I can come up with, but I honestly am not seeing your issue with `parents` sorry.

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: `includeParents` maybe?

[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/runner/Hooks.ts:20]: Yep that's better

File: modules/runner/src/main/ts/core/Test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/runner/src/main/ts/core/Test.ts:1]: I guess as I was reading through they looked really similar and then when comparing directly, you're right, it's mainly just the `fullTitle` function. I'll mark this as resolved

File: modules/runner/src/main/ts/core/Suite.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Suite.ts:18]: That still doesn't handle the implicit truthy check (yes we care about that). I also prefer to keep it appearing as though it's checking up the tree (both for performance and from how the logic actually operates).

File: modules/runner/src/main/ts/core/Test.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Test.ts:1]: Yeah, I admittedly was considering pulling that out. However, it was such a small bit of duplication I couldn't really figure where to put it and there was still the differences in the formatting (`/` vs `-`) so I just left it :shrug:

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/runner/src/main/ts/core/Test.ts:1]: Ended up pulling the `fullTitle` logic out into a separate util function.

**Review by metricjs — COMMENTED** [review]
I'm halfway, thought I'd submit these comments to speed things up. Apologies if I've commented on anything you explained earlier.

**Review by JamesToohey — APPROVED** [review]

File: modules/common/src/main/ts/api/TestTypes.ts [hunk 7]
```diff
@@ -0,0 +1,121 @@
+import { TestError } from './TestError';
+import { TestLabel } from './TestLabel';
+
+export const enum HookType {
+  Before = 'before',
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/TestTypes.ts:5]: It's what the BDD style uses... TDD is `beforeAll`.

File: modules/common/src/main/ts/api/Failure.ts [hunk 8]
```diff
@@ -30,6 +31,17 @@ export const normalizeError = (err: TestThrowable): TestError => {
...
     return normalizeError(err());
+  } else if (TestError.isAssertionError(err)) {
+    // Chai assertion errors are objects, but we want an actual error
+    // so it prints better in the console
+    const error = new Error(err.message) as AssertionError;
+    error.name = err.name;
+    error.stack = err.stack;
+    const keys = Object.keys(err) as (keyof AssertionError)[];
+    keys.forEach((key) => {
+      (error as any)[key] = err[key];
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/Failure.ts:42]: Yes, as `name` and `stack` are inherited properties (ie aren't in the list returned by `Object.keys`)

File: modules/common/src/main/ts/api/TestTypes.ts [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// eslint-disable-next-line @typescript-eslint/no-empty-interface
+export interface Hook extends Runnable {}
+
+export interface Test extends Runnable {
+  file?: string;
+  readonly parent?: Suite;
+  _only: boolean;
+
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/TestTypes.ts:88]: Yeah it was. I was trying to break up properties and methods/functions.

File: modules/common/src/main/ts/api/TestTypes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/TestTypes.ts:5]: Actually sorry I remembered incorrectly. Mocha uses `beforeAll` as a root-level hook that's used for global test setup: https://mochajs.org/#available-root-hooks. That hasn't been implemented at this stage, as I don't believe we'll ever need it (it's more useful in cli mode where you can define a prep script).

File: modules/common/src/main/ts/api/Reporter.ts [hunk 10]
```diff
@@ -94,14 +106,16 @@ ${dh}`;
 };
 
 const assertionErrorText = (e: AssertionError): string => {
-  const message = `Assertion error: ${e.message ? e.message : ''}
+  const actual = stringify(e.actual);
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/Reporter.ts:109]: Hmm... it's always been like this. I honestly like the clear separation between text and html, so I definitely don't want to lose that. I'll have a think, as the same applies to the `pprintAssertionErrorHtml` and `pprintAssertionErrorText` functions above this as well.

File: modules/common/src/main/ts/api/TestTypes.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/TestTypes.ts:5]: Also for reference, here's the general functions the BDD interface normally provides in mocha: https://mochajs.org/#bdd

[the remark this answers is not in the document]
Comment (JamesToohey) [reply modules/common/src/main/ts/api/TestTypes.ts:5]: Sweet 👍

File: modules/common/src/main/ts/api/Reporter.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/common/src/main/ts/api/Reporter.ts:80]: Have made a change to remove a fair bit of duplication. How does that look?

File: yarn.lock [hunk 11]
```diff
@@ -6402,7 +6402,7 @@ mkdirp@^0.5.0, mkdirp@^0.5.1, mkdirp@^0.5.3, mkdirp@^0.5.5:
   dependencies:
     minimist "^1.2.5"
 
-mocha@^8.2.0, mocha@^8.2.1:
+mocha@^8.2.0:
```
Comment (metricjs) [inline yarn.lock:6405]: Should mocha still be in here at all?

**Review by metricjs — APPROVED** [review]

File: yarn.lock (hunk 11 above)
Comment (lnewson) [reply yarn.lock:6405]: Yes, Mocha is still used to test all the bedrock functionality.

**Review by techtangents — APPROVED** [review]

File: modules/runner/package.json [hunk 12]
```diff
@@ -13,7 +13,6 @@
     "@ephox/bedrock-common": "^9.3.2",
     "@ephox/wrap-promise-polyfill": "^2.2.0",
     "jquery": "^3.4.1",
-    "mocha": "^8.2.1",
```
Comment (ashfordneil) [inline modules/runner/package.json:16]: Don't we still need this, just as a dev dependency? It looks like we still use mocha in the tests, it's just working due to this being listed as a dependency elsewhere in the repo.

**Review by ashfordneil — APPROVED** [review]

**Pull request merged**